sql-plugin/src/main/scala/org/apache/spark/sql/rapids/aggregate/GpuHLL.scala

```diff
@@ -0,0 +1,121 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids.aggregate
+
+import scala.collection.immutable.Seq
+
+import ai.rapids.cudf
+import ai.rapids.cudf.{DType, GroupByAggregation, ReductionAggregation}
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm.withResourceIfAllowed
+import com.nvidia.spark.rapids.RapidsPluginImplicits.ReallyAGpuExpression
+import com.nvidia.spark.rapids.jni.HLL
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression}
+import org.apache.spark.sql.catalyst.util.{GenericArrayData, HyperLogLogPlusPlusHelper}
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+
+case class CudfHLL(override val dataType: DataType,
+    numRegistersPerSketch: Int) extends CudfAggregate {
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (input: cudf.ColumnVector) => input.reduce(
+      ReductionAggregation.HLL(numRegistersPerSketch), DType.STRUCT)
+  override lazy val groupByAggregate: GroupByAggregation =
+    GroupByAggregation.HLL(numRegistersPerSketch)
+  override val name: String = "CudfHLL"
+}
+
+case class CudfMergeHLL(override val dataType: DataType,
+    numRegistersPerSketch: Int)
+    extends CudfAggregate {
+  override lazy val reductionAggregate: cudf.ColumnVector => cudf.Scalar =
+    (input: cudf.ColumnVector) =>
+      input.reduce(ReductionAggregation.mergeHLL(numRegistersPerSketch), DType.STRUCT)
+
+  override lazy val groupByAggregate: GroupByAggregation =
+    GroupByAggregation.mergeHLL(numRegistersPerSketch)
+  override val name: String = "CudfMergeHLL"
+}
+
+/**
+ * Perform the final evaluation step to compute approximate count distinct from sketches.
+ */
+case class GpuHLLEvaluation(childExpr: Expression, precision: Int)
+    extends GpuExpression with ShimExpression {
+  override def dataType: DataType = LongType
+
+  override def prettyName: String = "HLL_evaluation"
+
+  override def nullable: Boolean = false
+
+  override def children: Seq[Expression] = Seq(childExpr)
+
+  override def columnarEval(batch: ColumnarBatch): GpuColumnVector = {
+    withResourceIfAllowed(childExpr.columnarEval(batch)) { sketches =>
+      val distinctValues = HLL.estimateDistinctValueFromSketches(sketches.getBase, precision)
+      GpuColumnVector.from(distinctValues, LongType)
+    }
+  }
+}
+
+case class GpuHLL(childExpr: Expression, relativeSD: Double)
```

Comment:
Let' call by full name like `GpuHyperLogLogPlusPlus` to better reflect the CPU version.

Reply:
Done